src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs

```diff
@@ -308,11 +308,26 @@ private void CheckNullableAnnotationsInConstraints(DiagnosticBag diagnostics)
 
         internal bool ConstraintsNeedNullableAttribute()
         {
-            return (this.HasReferenceTypeConstraint && this.ReferenceTypeConstraintIsNullable != null) ||
-                   this.ConstraintTypesNoUseSiteDiagnostics.Any(c => c.NeedsNullableAttribute()) ||
-                   this.HasNotNullConstraint ||
-                   (!this.HasReferenceTypeConstraint && !this.HasValueTypeConstraint &&
-                    this.ConstraintTypesNoUseSiteDiagnostics.IsEmpty && this.IsNotNullableIfReferenceType == false);
+            if (!DeclaringCompilation.ShouldEmitNullableAttributes(ContainingSymbol))
```

Comment:
You might consider putting a link the nullable attribute spec here, so it's easier to understand the logic